File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """        
        # edge case: if one node only, kth smallest is that one node
        
            
        count = []
        
        def inorderSearch(node, count):
            if not node:
                return 
            
            inorderSearch(node.left, count)
            
            count.append(node.val)
            
            inorderSearch(node.right, count)
        
        inorderSearch(root,count)
        return count[k-1]
            
            

File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
    // recursive way
        // use in order traversal to get the tree nodes in a ordered list
        // and return arr[k - 1]
    
    // time: O(n), n = number of nodes in tree
    // space: O(n), n = number of nodes in tree
    const nums = []
    const dfs = function(root) {
        if (root == null) {
            return
        }
        
        dfs(root.left)
        nums.push(root.val)
        dfs(root.right)
    }
    
    dfs(root)
    return nums[k-1]
};